testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java

```diff
@@ -0,0 +1,268 @@
+   * not be used during the test teardown.
+   *
+   * @return the input server
+   */
+  public <T extends Server> T register(@Nonnull T server) {
+    checkNotNull(server, "server");
+    register(new ServerResource(server));
+    return server;
+  }
+
+  @VisibleForTesting
+  void register(Resource resource) {
+    resources.add(resource);
+  }
+
+  @Override
+  public Statement apply(final Statement base, Description description) {
+    return new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        try {
+          base.evaluate();
+        } catch (Throwable t) {
+          firstException = t;
+        } finally {
+          teardown();
+        }
+
+        if (firstException != null) {
+          throw firstException;
+        }
+      }
+    };
+  }
+
+  /**
+   * Releases all the registered resources.
+   */
+  private void teardown() {
+    stopwatch.start();
+
+    for (int i = resources.size() - 1; i >= 0; i--) {
+      if (firstException != null) {
+        break;
+      }
+
+      try {
+        resources.get(i).cleanUp();
+      } catch (Throwable t) {
+        if (firstException == null) {
+          firstException = t;
+        }
+        break;
+      }
+    }
+
+    for (int i = resources.size() - 1; i >= 0; i--) {
+      if (firstException != null) {
+        resources.get(i).forceCleanUp();
+        continue;
+      }
+
+      try {
+        boolean released = resources.get(i).awaitReleased(
+            timeoutNanos - stopwatch.elapsed(TimeUnit.NANOSECONDS), TimeUnit.NANOSECONDS);
+        if (!released) {
+          firstException = new AssertionError(
+              "Resource " + resources.get(i) + " can not be released in time at the end of test");
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        if (firstException == null) {
+          firstException = e;
+        }
+      } catch (Throwable t) {
+        if (firstException == null) {
+          firstException = t;
+        }
+      } finally {
+        if (firstException != null) {
+          resources.get(i).forceCleanUp();
+        }
+      }
+    }
+
+    resources.clear();
+  }
+
+  @VisibleForTesting
+  interface Resource {
+    void cleanUp();
+
+    /**
+     * Error already happened, try the best to clean up. Never throws.
+     */
+    void forceCleanUp();
+
+    /**
+     * Returns true if the resource is released in time.
+     */
+    boolean awaitReleased(long duration, TimeUnit timeUnit) throws InterruptedException;
+  }
+
+  private static final class ManagedChannelResource implements Resource {
+    final ManagedChannel channel;
+
+    ManagedChannelResource(ManagedChannel channel) {
+      this.channel = channel;
+    }
+
+    @Override
+    public void cleanUp() {
+      channel.shutdown();
+    }
+
+    @Override
+    public void forceCleanUp() {
+      try {
+        channel.shutdownNow();
+      } catch (Throwable ignorable) {
```

Comment:
shutdownNow should not throw. Let's not protect ourselves against that. If it happens, then we will just fix the bug.

Reply:
The javadoc didn't say that. Even `ManagedChannelImpl` behaves so and javadoc says so, the user's test could be registering any `ManagedChannel` to the rule, say a mock `ManagedChannel` delegating to a `ManagedChannelImpl`, and their mock could happen to be really wild and evil. Same for `awaitTermination()` and `shutdown()`.

Reply:
The methods do not define any exceptions will be thrown. Any Exception (vs Error) would be a bug (and as a general rule the amount of code that should avoid bugs in other code should be very limited) and any Error means "all hope is lost." If the user's test is doing something like that, I have no problem letting it fail. Those methods need to be reliable in general, and exposing such bugs seems fine to me. Garbage in, garbage out.

Reply:
Done.